chia/wallet/util/puzzle_decorator.py

```diff
@@ -0,0 +1,102 @@
+# General puzzle wrapper
+from __future__ import annotations
+
+from typing import Any, Dict, List, Tuple
+
+from typing_extensions import Protocol
+
+from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.wallet.puzzles.clawback.clawback_puzzle_decorator import ClawbackPuzzleDecorator
+from chia.wallet.util.puzzle_decorator_type import PuzzleDecoratorType
+
+
+class PuzzleDecoratorProtocol(Protocol):
+    # Protocol for puzzle decorators
+    @staticmethod
+    def create(config: Dict[str, Any]) -> PuzzleDecoratorProtocol:
+        ...
+
+    def decorate(self, inner_puzzle: Program) -> Program:
+        ...
+
+    def decorate_target_puzhash(self, inner_puzzle: Program, target_puzhash: bytes32) -> Tuple[Program, bytes32]:
+        ...
+
+    def decorate_memos(
+        self, inner_puzzle: Program, target_puzhash: bytes32, memos: List[bytes]
+    ) -> Tuple[Program, List[bytes]]:
+        ...
+
+    def solve(
+        self, puzzle: Program, primaries: List[Dict[str, Any]], inner_solution: Program
+    ) -> Tuple[Program, Program]:
+        ...
+
+
+class PuzzleDecoratorManager:
+    decorator_list: List[PuzzleDecoratorProtocol]
```

Comment:
For the Clawback there is nothing to worry about since we don't decorate the puzzle.
 - It's like a pipeline. The decorators will be applied one by one. The order does matter here.
 - It depends on if the decorator will change the puzzle. If it does, then the user will not see their coins, but they still own the coins as long as they can recover the config. If it doesn't change the puzzle then no impact on the user basically. It just changed how they will spend their coin.

Reply:
Hm so this means for a decorated puzzle, if the user loses its config or wants to use this coins in another wallet it's not possible to recover it from the chain and the funds depend not only on the mnemonic anymore. It feels like we should add some metadata to the coin, on chain, so that we can spend coins even if the config was changed, accidentally removed or lost? Maybe i misunderstand the concept here and it's not possible to avoid this for this feature but it sounds wrong to have funds which depend on the config file, no?